src/OracleService/OracleService.cs

```diff
@@ -0,0 +1,476 @@
+            {
+                var result = code == OracleResponseCode.Success ? Filter(data, request.Filter) : Array.Empty<byte>();
+                var response = new OracleResponse() { Id = requestId, Code = code, Result = result };
+                var responseTx = CreateResponseTx(snapshot, response);
+                var backupTx = CreateResponseTx(snapshot, new OracleResponse() { Code = OracleResponseCode.ConsensusUnreachable, Id = requestId, Result = Array.Empty<byte>() });
+
+                Log($"Builded response tx:{responseTx.Hash} requestTx:{request.OriginalTxid} requestId: {requestId}");
+
+                List<Task> tasks = new List<Task>();
+                ECPoint[] oraclePublicKeys = NativeContract.RoleManagement.GetDesignatedByRole(snapshot, Role.Oracle, snapshot.Height + 1);
+                foreach (var account in wallet.GetAccounts())
+                {
+                    var oraclePub = account.GetKey().PublicKey;
+                    if (!account.HasKey || account.Lock || !oraclePublicKeys.Contains(oraclePub)) continue;
+
+                    var txSign = responseTx.Sign(account.GetKey());
+                    var backTxSign = backupTx.Sign(account.GetKey());
+                    AddResponseTxSign(snapshot, requestId, oraclePub, txSign, responseTx, backupTx, backTxSign);
+                    tasks.Add(SendResponseSignatureAsync(requestId, txSign, account.GetKey()));
+
+                    Log($"Send oracle sign data: Oracle node: {oraclePub} RequestTx: {request.OriginalTxid} Sign: {txSign.ToHexString()}");
+                }
+                await Task.WhenAll(tasks);
+            }
+        }
+
+        private async void ProcessRequestsAsync()
+        {
+            while (!cancelSource.IsCancellationRequested)
+            {
+                using (var snapshot = Blockchain.Singleton.GetSnapshot())
+                {
+                    foreach (var (id, request) in NativeContract.Oracle.GetRequests(snapshot))
+                    {
+                        if (cancelSource.IsCancellationRequested) break;
+                        if (!finishedCache.ContainsKey(id) && (!pendingQueue.TryGetValue(id, out OracleTask task) || task.Tx is null))
+                            await ProcessRequestAsync(snapshot, request);
+                    }
+                }
+                if (cancelSource.IsCancellationRequested) break;
+                await Task.Delay(500);
+            }
+            stoped = true;
+        }
+
+        private async Task<(OracleResponseCode, string)> ProcessUrlAsync(string url)
+        {
+            if (!Uri.TryCreate(url, UriKind.Absolute, out var uri))
+                return (OracleResponseCode.Error, null);
+            if (!protocols.TryGetValue(uri.Scheme, out IOracleProtocol protocol))
+                return (OracleResponseCode.ProtocolNotSupported, null);
+            try
+            {
+                return await protocol.ProcessAsync(uri, cancelSource.Token);
```

Comment:
protocol use `private readonly HttpClient client = new HttpClient();` that I think that it's not thread safe, I think that we should create the protocol for each request.

Reply:
It's thread safe.